qutip/lattice.py

```diff
@@ -0,0 +1,814 @@
+                        col_ind = np.append(col_ind, [(lin_RI-nx_units)*NS+l])
+                        data = np.append(data, [np.conj(inter_hop_y[l, k])])
+
+            for k in range(x1):
+                for l in range(y1):
+                    if (j < (ny_units-1)):
+                        row_ind = np.append(row_ind, [lin_RI*NS+k])
+                        col_ind = np.append(col_ind, [(lin_RI+nx_units)*NS+l])
+                        data = np.append(data, [inter_hop_y[k, l]])
+
+    M = nx_units * ny_units * NS
+
+    for i in range(0, nx_units):
+        lin_RI = i
+        if (PBCy == 1 and ny_units * length_for_site > 2):
+            for k in range(x1):
+                for l in range(y1):
+                    (Aw,) = np.where(row_ind == (lin_RI * NS+k))
+                    (Bw,) = np.where(col_ind == ((lin_RI+(ny_units-1) *
+                                                 nx_units) * NS+l))
+                    if (len(np.intersect1d(Aw, Bw)) == 0):
+                        row_ind = np.append(
+                                row_ind,
+                                [lin_RI * NS+k,
+                                 (lin_RI+(ny_units-1) * nx_units) * NS+l])
+
+                        col_ind = np.append(
+                                col_ind, [(lin_RI +
+                                          (ny_units-1) * nx_units) * NS+l,
+                                          lin_RI * NS+k])
+
+                        data = np.append(data, [np.conj(inter_hop_y[l, k]),
+                                                inter_hop_y[l, k]])
+
+    for j in range(0, ny_units):
+        lin_RI = j
+        if (PBCx == 1 and nx_units * cell_num_site > 2):
+            for k in range(x0):
+                for l in range(y0):
+                    (Aw,) = np.where(row_ind == (lin_RI * nx_units * NS+k))
+                    (Bw,) = np.where(col_ind ==
+                                     (((lin_RI+1) * nx_units-1) * NS+l))
+                    if (len(np.intersect1d(Aw, Bw)) == 0):
+                        for m in range(xi_len):
+                            row_ind = np.append(
+                                    row_ind, [lin_RI * nx_units * NS+k,
+                                              ((lin_RI+1) * nx_units-1) *
+                                              NS+l])
+
+                            col_ind = np.append(col_ind, [((lin_RI+1) *
+                                                           nx_units-1) * NS+l,
+                                                          lin_RI *
+                                                          nx_units * NS+k])
+
+                            data = np.append(
+                                    data, [np.conj(inter_hop_x[m][l, k]),
+                                           inter_hop_x[m][l, k]])
+
+    Hamt = csr_matrix((data, (row_ind, col_ind)), [M, M], dtype=np.complex)
+    return Hamt
+
+
+def diag_a_matrix(H_k, calc_evecs=False):
```

Comment:
What is the reason for not using QuTiP's `eigenstates` and `eigenenergies` functions?